brainglobe_registration/similarity_metrics_1.py

```diff
@@ -0,0 +1,427 @@
+        return 0.0  # return a very low score if there's no signal
+    return np.corrcoef(img1.ravel(), img2.ravel())[0, 1]
+
+
+def compute_similarity_metric(moving, fixed, atlas_res=(1, 1, 1)):
+    # moving = moving.astype(np.float32)
+    # fixed = fixed.astype(np.float32)
+
+    moving, fixed = prepare_images(moving, fixed)
+
+    # Similarity metrics
+    ncc = safe_ncc(moving, fixed)
+
+    # Structural Similarity Index
+    # Captures luminance, contrast, structure
+    ssim = structural_similarity(moving, fixed, data_range=1.0)
+
+    # Mutual Information
+    # Shared info, robust to intensity + spatial differences
+    mi = mutual_info_regression(moving.ravel().reshape(-1, 1), fixed.ravel())[
+        0
+    ]
+
+    combined = 0.55 * ssim + 0.35 * ncc + 0.1 * mi
+    return combined
+
+
+def create_rotation_matrix(roll: float, yaw: float, pitch: float):
+    """Create a combined 3D rotation matrix from roll, yaw,
+    and pitch (in degrees)."""
+    # Create the rotation matrix
+    roll_matrix = active_matrix_from_angle(0, np.deg2rad(roll))
+    yaw_matrix = active_matrix_from_angle(1, np.deg2rad(yaw))
+    pitch_matrix = active_matrix_from_angle(2, np.deg2rad(pitch))
+
+    # Combine rotation matrices
+    rotation_matrix = yaw_matrix @ pitch_matrix @ roll_matrix
+
+    full_matrix = np.eye(4)
+    full_matrix[:3, :3] = rotation_matrix
+    return full_matrix
+
+
+def rotate_volume(
+    data: np.ndarray,
+    rotation_matrix: np.ndarray,
+    reference_shape: Tuple[int, int, int],
+    interpolation_order: int = 2,
+):
+    """Rotate a 3D volume using a given rotation matrix and return
+    transformed data and transform matrix."""
+
+    # Translate the origin to the center of the image
+    origin = np.asarray(reference_shape) / 2
+
+    translate_to_center = np.eye(4)
+    translate_to_center[:3, -1] = -origin
+
+    bounding_box = calculate_rotated_bounding_box(
+        reference_shape, rotation_matrix
+    )
+    new_translation = np.asarray(bounding_box) / 2
+
+    post_rotate_translation = np.eye(4)
+    post_rotate_translation[:3, -1] = new_translation
+
+    # Combine the matrices. The order of operations is:
+    # 1. Translate the origin to the center of the image
+    # 2. Rotate the image
+    # 3. Translate the origin back to the top left corner
+
+    final_transform = np.linalg.inv(
+        post_rotate_translation @ rotation_matrix @ translate_to_center
+    )
+
+    transformed = ndi.affine_transform(
+        da.from_array(
+            data, chunks=(2, reference_shape[1], reference_shape[2])
+        ),
+        matrix=final_transform,
+        output_shape=bounding_box,
+        order=interpolation_order,
+    ).astype(data.dtype)
+
+    return transformed, final_transform, bounding_box
+
+
+# Optimisation objective
+# ----------------------------------------
+# Load data
+# ----------------------------------------
+atlas_name = (get_downloaded_atlases())[0]  # e.g. Allen 100um
```

Comment:
I would just refer to the atlas by name here while experimenting
```suggestion
atlas_name ="allen_mouse_100um"
```